### PR TITLE
show the right image name in job log and docker ps

### PR DESCRIPTION
**Pull request by liusdu**
### introduction

It is a new version of PR #10479 . 

> When we tag an Image with several names and we run one of them,
> The "create" job will log this event with
> +job log(create, containerID, Imagename).
> And the "Imagename" is always the first one (sorted). It is the
> same to "start/stop/rm" jobs. So use the correct name instand.

 I have discuss this with @LK4D4. We all think we should show just as what we type. He afraid that this may introduce break change. @tiborvass thinks we can show ubuntu:latest in the first case and 8eaa4ff0 in the second case. 

But  because when we start a container with Image ID. We cannot record a related image:tag in log.
So this will introduce a break change(not image:tag format).  So my patched does following.
### Image List

---

| Image tag | ImageID |
| --- | --- |
| a-ubuntu:latest | 8eaa4ff0 |
| ubuntu:latest | 8eaa4ff0 |
| z-ubunutu:latest | 8eaa4ff0 |
### behavior before  this patch series

---

| No | cmd | behaviour |
| --- | --- | --- |
| 1 | docker run -ti ubuntu bash | +jog(create, containerid, a-ubuntu:latest) |
| 3 | docker run -ti ubuntu:latest bash | +jog(create, containerid, a-ubuntu:latest) |
| 2 | docker run -ti 8eaa4ff0 bash | +jog(create, containerid, a-ubuntu:latest) |

---
### behavior  after this patch series

---

| No | cmd | behaviour |
| --- | --- | --- |
| 1 | docker run -ti ubuntu bash | +jog(create, containerid, ubuntu) |
| 3 | docker run -ti ubuntu:latest bash | +jog(create, containerid, ubuntu:latest) |
| 2 | docker run -ti 8eaa4ff0 bash | +jog(create, containerid, 8eaa4ff0) |

---
### extra changes
- Because the daemon log behavior has changed. It affects "docker events", which was used several times in ci. So I also fix these issues.
- docker ps also has similar issues. Fix them

### CODE REVIEW

**Comment by LK4D4** [comment]
I'm okay with this.
Could you pls squash fixup commits?

**Comment by liusdu** [comment]
hi @LK4D4 done， Please help me review

File: daemon/list.go [hunk 1]
```diff
@@ -130,10 +127,6 @@ func (daemon *Daemon) Containers(job *engine.Job) error {
 		out.SetJson("Id", container.ID)
 		out.SetList("Names", names[container.ID])
 		img := container.Config.Image
-		_, tag := parsers.ParseRepositoryTag(container.Config.Image)
-		if tag == "" {
-			img = utils.ImageReference(img, graph.DEFAULTTAG)
-		}
 		out.SetJson("Image", img)
```
Comment (LK4D4) [inline daemon/list.go]: can be just `container.Config.Image` now.

File: integration-cli/docker_cli_daemon_test.go [hunk 2]
```diff
@@ -292,6 +293,60 @@ func TestDaemonLoggingLevel(t *testing.T) {
 	logDone("daemon - Logging Level")
 }
 
+func TestDaemonLoggingContent(t *testing.T) {
+
```
Comment (LK4D4) [inline integration-cli/docker_cli_daemon_test.go]: Would you mind to remove empty lines after `{` and before `}`?

**Comment by tiborvass** [comment]
@liusdu so you're basically agreeing with my comment https://github.com/docker/docker/pull/10479#issuecomment-81958550 ? Just wanna make sure :)

File: integration-cli/docker_cli_daemon_test.go (hunk 2 above)
Comment (liusdu) [inline integration-cli/docker_cli_daemon_test.go]: Thanks ! @LK4D4

**Comment by liusdu** [comment]
@tiborvass Thanks to you reply. I am afraid of this scenario: one starts two containers with busybox and busybox:latest.  With this patch docker events can identify them using flag "--filter".  But if we add default tag for busybox, we can not do it.

**Comment by LK4D4** [comment]
LGTM

File: integration-cli/docker_cli_daemon_test.go [hunk 3]
```diff
@@ -292,6 +293,58 @@ func TestDaemonLoggingLevel(t *testing.T) {
 	logDone("daemon - Logging Level")
 }
 
+func TestDaemonLoggingContent(t *testing.T) {
+	CheckImageNameInLog := func(d *Daemon, ImageName string) {
```
Comment (tiborvass) [inline integration-cli/docker_cli_daemon_test.go]: should start with small letters: `checkImageNameInLog` and `imageName`

File: integration-cli/docker_cli_daemon_test.go [hunk 4]
```diff
@@ -292,6 +293,58 @@ func TestDaemonLoggingLevel(t *testing.T) {
...
+
+		out, err := d.Cmd("run", "-d", ImageName)
+		if err != nil {
+			t.Fatal(err, out)
+		}
+		content, _ := ioutil.ReadFile(d.logFile.Name())
+
+		reg := regexp.MustCompile(`\+job log\(create, (?P<ID>[0-9|a-z]*), (?P<ImageName>.*)\)`)
```
Comment (tiborvass) [inline integration-cli/docker_cli_daemon_test.go]: you should be able to declare this outside the function

File: integration-cli/docker_cli_daemon_test.go [hunk 5]
```diff
@@ -292,6 +293,58 @@ func TestDaemonLoggingLevel(t *testing.T) {
...
+		if Names == nil {
+			t.Fatalf("Could not find create logs")
+		}
+		if Names[2] != ImageName {
+			t.Fatalf("Wrong Image name:\nexpect:\t+job log(create, %s, %s)\ngot:\t+job log(create, %s, %s)", Names[1], ImageName, Names[1], Names[2])
+		}
+	}
+
+	// check Image name in job log
+	ImageNames := []string{"a-busybox", "b-busybox", "c-busybox"}
```
Comment (tiborvass) [inline integration-cli/docker_cli_daemon_test.go]: `imageNames`

**Comment by tiborvass** [comment]
@liusdu couple of style changes, and it's good otherwise :)

**Comment by liusdu** [comment]
Hi,  @tiborvass all done.

**Comment by liusdu** [comment]
I add a bad patch to let event test pass .
in commit 6b5e2088 I sleep one second to let container stops. It is very bad.

But now I have a better idea. I can directly stop containers,not just wait. So related tests should be changed acorrdingly.

**Comment by liusdu** [comment]
Pr #11981 also encounters this problem.

**Comment by LK4D4** [comment]
@liusdu It is fixed in #11949 (I hope at least).

**Comment by tiborvass** [comment]
@LK4D4 I can confirm it was not fixed by #11949. I still see `busybox:latest` in all cases.

**Comment by tiborvass** [comment]
LGTM

**Comment by LK4D4** [comment]
@tiborvass I meant event tests fail.

**Comment by LK4D4** [comment]
@liusdu Could you remove your patch to events test and rebase against master?

**Comment by liusdu** [comment]
@LK4D4  @tiborvass  done

File: integration-cli/docker_cli_daemon_test.go [hunk 6]
```diff
@@ -291,6 +292,57 @@ func TestDaemonLoggingLevel(t *testing.T) {
 
 	logDone("daemon - Logging Level")
 }
+func checkImageNameInLog(d *Daemon, imageName string) (error, string) {
```
Comment (tiborvass) [inline integration-cli/docker_cli_daemon_test.go]: @liusdu sorry you should keep this inside the test function like so:

```Go
var checkImageNameInLog = func(...) (...) {
    // ...
}
```

Comment (LK4D4) [inline integration-cli/docker_cli_daemon_test.go]: or even:
```
checkImageNameInLog := func(...) (...) {
    // ...
}
```

**Comment by liusdu** [comment]
@LK4D4  @tiborvass  Good advice, Thank you! done

File: integration-cli/docker_cli_daemon_test.go [hunk 7]
```diff
@@ -292,6 +293,57 @@ func TestDaemonLoggingLevel(t *testing.T) {
...
+	}
+	ImageID := stripTrailingCharacters(out)
+	testImages := []string{imageNames[1], ImageID}
+
+	for _, testImage := range testImages {
+		if err, out := checkImageNameInLog(d, testImage); err != nil {
+			t.Fatal(err, out)
+		}
+	}
+	d.Stop()
```
Comment (jfrazelle) [inline integration-cli/docker_cli_daemon_test.go]: can we defer d.Stop() on line 324

Comment (liusdu) [inline integration-cli/docker_cli_daemon_test.go]: Sorry for my bad code style!

**Comment by LK4D4** [comment]
LGTM

**Comment by jfrazelle** [comment]
ah man needs a rebase so sorry

**Comment by liusdu** [comment]
@jfrazelle  all fixed!

**Comment by jfrazelle** [comment]
I dont know if that worked, it's still saying unable to merge due to conflicts? was i too slow :(

**Comment by liusdu** [comment]
@jfrazelle  Yesterday I fixed conflicts with two commits. Today one more~~.  Codes related with this patch is in heave developing.  And after commit c9eb37f9., we do not the test for daemon log.

**Comment by jfrazelle** [comment]
LGTM so sorry!

**Comment by LK4D4** [comment]
Ah, nice!

**Comment by relistan** [comment]
This change unfortunately broke `docker ps` output as shown in #12595. The breaking change is here: https://github.com/liusdu/docker/commit/645c020f5ab7119cd06d7c6a790e9d99fe1cd309#diff-9320049615863c9b6ab00c749bbb2f09L139

**Pull request merged**